Konnectivity dockerfile fix

makhov opened the pull request:
**What this PR Includes**
The PR fixes docker build error:
```
#9 38.66 make: protoc: No such file or directory
```

Konnectivity's readme says:
> Proto definitions are compiled with protoc. Please ensure you have protoc installed (Instructions) and the proto-gen-go library at the appropriate version.

https://github.com/kubernetes-sigs/apiserver-network-proxy#protoc


Full log:
```
docker build -t k0sbuild.docker-image.konnectivity \
                --build-arg VERSION=0.0.24 \
                --build-arg BUILDIMAGE=golang:1.16-alpine \
                --build-arg BUILD_GO_TAGS= \
                --build-arg BUILD_GO_CGO_ENABLED=0 \
                --build-arg BUILD_SHIM_GO_CGO_ENABLED= \
                --build-arg BUILD_GO_FLAGS="-a" \
                --build-arg BUILD_GO_LDFLAGS="-w -s" \
                --build-arg BUILD_GO_LDFLAGS_EXTRA="-extldflags=-static" \
                -f konnectivity/Dockerfile .
[+] Building 39.3s (9/10)                                                                                                                                                                                 
 => [internal] load build definition from Dockerfile                                                                                                                                                 0.0s
 => => transferring dockerfile: 904B                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                    0.0s
 => => transferring context: 34B                                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/golang:1.16-alpine                                                                                                                                0.2s
 => [build 1/6] FROM docker.io/library/golang:1.16-alpine@sha256:69d613a986635e0ee3f1b351131fa689b0dd3570ef9cab8e619a5f1fb939ec56                                                                    0.0s
 => CACHED [build 2/6] RUN apk add build-base git make                                                                                                                                               0.0s
 => CACHED [build 3/6] RUN git clone -b v0.0.24 --depth=1 https://github.com/kubernetes-sigs/apiserver-network-proxy.git /apiserver-network-proxy                                                    0.0s
 => CACHED [build 4/6] WORKDIR /apiserver-network-proxy                                                                                                                                              0.0s
 => CACHED [build 5/6] RUN go version                                                                                                                                                                0.0s
 => ERROR [build 6/6] RUN GO111MODULE=on go get github.com/golang/mock/mockgen@v1.4.4 &&     make gen &&     CGO_ENABLED=0     GOOS=linux     go build         -a         -tags=""         -ldflag  39.0s
------                                                                                                                                                                                                    
 > [build 6/6] RUN GO111MODULE=on go get github.com/golang/mock/mockgen@v1.4.4 &&     make gen &&     CGO_ENABLED=0     GOOS=linux     go build         -a         -tags=""         -ldflags="-w -s -extldflags=-static"         -o bin/proxy-server cmd/server/main.go:                                                                                                                                            
#9 3.051 go: downloading github.com/golang/mock v1.4.4
#9 3.180 go: downloading golang.org/x/tools v0.0.0-20210106214847-113979e3529a
#9 4.545 go: downloading golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1
#9 4.545 go: downloading golang.org/x/mod v0.3.0
#9 6.958 go mod download
#9 38.66 mkdir -p /go/src
#9 38.66 protoc -I . proto/agent/agent.proto --go_out=plugins=grpc:/go/src
#9 38.66 make: protoc: No such file or directory
#9 38.66 make: *** [Makefile:110: proto/agent/agent.pb.go] Error 127
------
executor failed running [/bin/sh -c GO111MODULE=on go get github.com/golang/mock/mockgen@v1.4.4 &&     make gen &&     CGO_ENABLED=${BUILD_GO_CGO_ENABLED}     GOOS=linux     go build         ${BUILD_GO_FLAGS}         -tags="${BUILD_GO_TAGS}"         -ldflags="${BUILD_GO_LDFLAGS} ${BUILD_GO_LDFLAGS_EXTRA}"         -o bin/proxy-server cmd/server/main.go]: exit code: 2
make[1]: *** [.docker-image.konnectivity.stamp] Error 1
make: *** [.bins.linux.stamp] Error 2

```

makhov commented:
Hm, it turned out that I had an old `golang:1.16-alpine` image. 
`docker pull golang:1.16-alpine` fixed the issue

xinfengliu commented:
In my environment, there's no need to install `protoc` for building konnectivity. Maybe you can try `make clean` then `make build`?

```log
#10 [build 6/6] RUN GO111MODULE=on go get github.com/golang/mock/mockgen@v1.4.4 &&     make gen &&     CGO_ENABLED=0     GOOS=linux     go build         -a         -tags=""         -ldflags="-w -s -extldflags=-static"         -o bin/proxy-server cmd/server/main.go
#10 sha256:9e65a33d11e1f0be88cfc1b8ea82f45c6c8f3db8b057e2b86b1a46ebf2d65030
#10 1.350 go: downloading github.com/golang/mock v1.4.4
#10 1.390 go: downloading golang.org/x/tools v0.0.0-20210106214847-113979e3529a
#10 1.810 go: downloading golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1
#10 1.810 go: downloading golang.org/x/mod v0.3.0
#10 2.079 go get: installing executables with 'go get' in module mode is deprecated.
#10 2.079       To adjust and download dependencies of the current module, use 'go get -d'.
#10 2.079       To install using requirements of the current module, use 'go install'.
#10 2.079       To install ignoring the current module, use 'go install' with a version,
#10 2.079       like 'go install example.com/cmd@latest'.
#10 2.079       For more information, see https://golang.org/doc/go-get-install-deprecation
#10 2.079       or run 'go help get' or 'go help install'.
#10 3.518 go mod download
#10 9.486 mkdir -p proto/agent/mocks
#10 9.487 mockgen sigs.k8s.io/apiserver-network-proxy/proto/agent AgentService_ConnectServer > proto/agent/mocks/agent_mock.go
#10 15.11 cat hack/go-license-header.txt proto/agent/mocks/agent_mock.go > proto/agent/mocks/agent_mock.licensed.go
#10 15.12 mv proto/agent/mocks/agent_mock.licensed.go proto/agent/mocks/agent_mock.go
#10 DONE 59.1s
```

makhov commented:
How does it work? We need to have `protoc` installed to run `make gen` and it's not pre-installed in `golang:1.16-alpine`. So I had to install it explicitly.

I've just tried and it still doesn't work:
```
❯ make clean && make build                 
docker volume rm k0sbuild.gocache
Error: No such volume: k0sbuild.gocache
make: [clean-gocache] Error 1 (ignored)
rm .k0sbuild.docker-vol.gocache
rm: .k0sbuild.docker-vol.gocache: No such file or directory
make: [clean-gocache] Error 1 (ignored)
docker rmi k0sbuild.docker-image.k0s -f
Error: No such image: k0sbuild.docker-image.k0s
rm -f .k0sbuild.docker-image.k0s
rm -f pkg/assets/zz_generated_offsets_*.go k0s k0s.exe .bins.*stamp bindata* static/gen_manifests.go
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C embedded-bins clean
for i in .container.*; do \
                if [ -f $i ]; then \
                        docker rm $(cat $i) && rm $i; \
                fi; \
        done
for i in .docker-image.*; do \
                if [ -f $i ]; then \
                        docker rmi -f k0sbuild$(basename $i .stamp) && rm $i;\
                fi; \
        done
Untagged: k0sbuild.docker-image.containerd:latest
Deleted: sha256:a37d848bfb323696b0ee7e0aa273344f682f2fc03c3985624d9c3ccb8c62d1c0
Untagged: k0sbuild.docker-image.etcd:latest
Deleted: sha256:150cc534ae66817e609527621c6eea2ffe9ae511edbf28fbf4a1fb413db4e910
Untagged: k0sbuild.docker-image.kine:latest
Deleted: sha256:bec37dd062d0d13dd6d6105f3fc24d12e12226abfcda3bc8286c8c3473d6771e
Untagged: k0sbuild.docker-image.kubernetes:latest
Deleted: sha256:fb0ae288c08246361af879fc27ba6a1dbce1ae9dd7cada346ecb4e574599802e
Untagged: k0sbuild.docker-image.runc:latest
Deleted: sha256:e8a804ed5f4b7c522a907baae55d9f9df541e69193eadd659dc7aaa9276671ac
rm -rf staging .tmp/* *.tmp
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C image-bundle clean
rm -f bundle.tar
rm -f .docker-bundler.stamp
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C inttest clean
rm -rf bin sonobuoy/*_sonobuoy_*.tar.gz .*.stamp
docker build --rm -t k0sbuild.docker-image.k0s -f build/Dockerfile .
[+] Building 3.3s (7/7) FINISHED                                                                                                                                                                          
 => [internal] load build definition from Dockerfile                                                                                                                                                 0.0s
 => => transferring dockerfile: 117B                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                    0.0s
 => => transferring context: 72B                                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/golang:1.16-alpine                                                                                                                                3.1s
 => [auth] library/golang:pull token for registry-1.docker.io                                                                                                                                        0.0s
 => [1/2] FROM docker.io/library/golang:1.16-alpine@sha256:69d613a986635e0ee3f1b351131fa689b0dd3570ef9cab8e619a5f1fb939ec56                                                                          0.0s
 => CACHED [2/2] RUN apk add --no-cache gcc musl-dev binutils-gold                                                                                                                                   0.0s
 => exporting to image                                                                                                                                                                               0.0s
 => => exporting layers                                                                                                                                                                              0.0s
 => => writing image sha256:8850ad19ce17263d5227662ce31e003313d4d8a36f7a8b392e778105527490dc                                                                                                         0.0s
 => => naming to docker.io/library/k0sbuild.docker-image.k0s                                                                                                                                         0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
touch .k0sbuild.docker-image.k0s
docker volume create k0sbuild.gocache
k0sbuild.gocache
docker run --rm -v k0sbuild.gocache:/gocache alpine:latest install -d -o $(id -u) -g $(id -g) /gocache/mod /gocache/build
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
a0d0a0d46f8b: Already exists 
Digest: sha256:e1c082e3d3c45cccac829840a25941e679c25d438cc8412c2fa221cf1a824e6a
Status: Downloaded newer image for alpine:latest
touch .k0sbuild.docker-vol.gocache
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C embedded-bins buildmode=docker TARGET_OS=linux
mkdir -p staging/linux/bin
docker build -t k0sbuild.docker-image.runc \
                --build-arg VERSION=1.0.2 \
                --build-arg BUILDIMAGE=golang:1.16-alpine \
                --build-arg BUILD_GO_TAGS="seccomp" \
                --build-arg BUILD_GO_CGO_ENABLED= \
                --build-arg BUILD_SHIM_GO_CGO_ENABLED= \
                --build-arg BUILD_GO_FLAGS= \
                --build-arg BUILD_GO_LDFLAGS= \
                --build-arg BUILD_GO_LDFLAGS_EXTRA="-w -s -extldflags=-static" \
                -f runc/Dockerfile .
[+] Building 0.9s (17/17) FINISHED                                                                                                                                                                        
 => [internal] load build definition from Dockerfile                                                                                                                                                 0.0s
 => => transferring dockerfile: 1.25kB                                                                                                                                                               0.0s
 => [internal] load .dockerignore                                                                                                                                                                    0.0s
 => => transferring context: 49B                                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/golang:1.16-alpine                                                                                                                                0.7s
 => [build  1/12] FROM docker.io/library/golang:1.16-alpine@sha256:69d613a986635e0ee3f1b351131fa689b0dd3570ef9cab8e619a5f1fb939ec56                                                                  0.0s
 => CACHED [build  2/12] RUN apk add build-base git  curl linux-headers gperf bash pkgconf                                                                                                           0.0s
 => CACHED [build  3/12] RUN curl -L https://github.com/seccomp/libseccomp/releases/download/v2.5.1/libseccomp-2.5.1.tar.gz  | tar -C / -zx                                                          0.0s
 => CACHED [build  4/12] RUN cd /libseccomp-2.5.1 && ./configure --sysconfdir=/etc --enable-static                                                                                                   0.0s
 => CACHED [build  5/12] RUN make -j$(nproc) -C /libseccomp-2.5.1                                                                                                                                    0.0s
 => CACHED [build  6/12] RUN make -j$(nproc) -C /libseccomp-2.5.1 check                                                                                                                              0.0s
 => CACHED [build  7/12] RUN make -C /libseccomp-2.5.1 install                                                                                                                                       0.0s
 => CACHED [build  8/12] RUN mkdir -p /go/src/github.com/opencontainers/runc                                                                                                                         0.0s
 => CACHED [build  9/12] RUN git clone -b v1.0.2 --depth=1 https://github.com/opencontainers/runc.git /go/src/github.com/opencontainers/runc                                                         0.0s
 => CACHED [build 10/12] WORKDIR /go/src/github.com/opencontainers/runc                                                                                                                              0.0s
 => CACHED [build 11/12] RUN go version                                                                                                                                                              0.0s
 => CACHED [build 12/12] RUN make  CGO_ENABLED=  BUILDTAGS="seccomp"  EXTRA_FLAGS=""  EXTRA_LDFLAGS="-w -s -extldflags=-static"                                                                      0.0s
 => CACHED [stage-1 1/1] COPY --from=build /go/src/github.com/opencontainers/runc/runc /bin/runc                                                                                                     0.0s
 => exporting to image                                                                                                                                                                               0.0s
 => => exporting layers                                                                                                                                                                              0.0s
 => => writing image sha256:e8a804ed5f4b7c522a907baae55d9f9df541e69193eadd659dc7aaa9276671ac                                                                                                         0.0s
 => => naming to docker.io/library/k0sbuild.docker-image.runc                                                                                                                                        0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
touch .docker-image.runc.stamp
docker create k0sbuild.docker-image.runc > .container.runc.tmp
mv .container.runc.tmp .container.runc
docker export $(cat .container.runc) | tar -C staging/linux/ -xv bin/runc && touch staging/linux/bin/runc
tar: Failed to set default locale
x bin/runc
docker build -t k0sbuild.docker-image.kubernetes \
                --build-arg VERSION=1.22.2 \
                --build-arg BUILDIMAGE=golang:1.16-alpine \
                --build-arg BUILD_GO_TAGS="providerless" \
                --build-arg BUILD_GO_CGO_ENABLED= \
                --build-arg BUILD_SHIM_GO_CGO_ENABLED= \
                --build-arg BUILD_GO_FLAGS="-v" \
                --build-arg BUILD_GO_LDFLAGS= \
                --build-arg BUILD_GO_LDFLAGS_EXTRA="-w -s -extldflags=-static" \
[+] Building 0.4s (12/12) FINISHED                                                                                                                                                                        
 => [internal] load build definition from Dockerfile                                                                                                                                                 0.0s
 => => transferring dockerfile: 1.42kB                                                                                                                                                               0.0s
 => [internal] load .dockerignore                                                                                                                                                                    0.0s
 => => transferring context: 34B                                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/golang:1.16-alpine                                                                                                                                0.2s
 => [build 1/7] FROM docker.io/library/golang:1.16-alpine@sha256:69d613a986635e0ee3f1b351131fa689b0dd3570ef9cab8e619a5f1fb939ec56                                                                    0.0s
 => CACHED [build 2/7] RUN apk add build-base git go-bindata linux-headers rsync grep coreutils bash                                                                                                 0.0s
 => CACHED [build 3/7] RUN mkdir -p /go/src/github.com/kubernetes/kubernetes                                                                                                                         0.0s
 => CACHED [build 4/7] RUN git clone -b v1.22.2 --depth=1 https://github.com/kubernetes/kubernetes.git /go/src/github.com/kubernetes/kubernetes                                                      0.0s
 => CACHED [build 5/7] WORKDIR /go/src/github.com/kubernetes/kubernetes                                                                                                                              0.0s
 => CACHED [build 6/7] RUN go version                                                                                                                                                                0.0s
 => CACHED [build 7/7] RUN  if [ 0 -eq 1 ]; then   export KUBE_CGO_OVERRIDES="kubelet kube-apiserver kube-scheduler kube-controller-manager";  fi;  for cmd in kubelet kube-apiserver kube-schedule  0.0s
 => CACHED [stage-1 1/1] COPY --from=build  /go/src/github.com/kubernetes/kubernetes/_output/local/bin/linux/*/kubelet  /go/src/github.com/kubernetes/kubernetes/_output/local/bin/linux/*/kube-api  0.0s
 => exporting to image                                                                                                                                                                               0.0s
 => => exporting layers                                                                                                                                                                              0.0s
 => => writing image sha256:fb0ae288c08246361af879fc27ba6a1dbce1ae9dd7cada346ecb4e574599802e                                                                                                         0.0s
 => => naming to docker.io/library/k0sbuild.docker-image.kubernetes                                                                                                                                  0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
touch .docker-image.kubernetes.stamp
docker create k0sbuild.docker-image.kubernetes > .container.kubernetes.tmp
mv .container.kubernetes.tmp .container.kubernetes
docker export $(cat .container.kubernetes) | tar -C staging/linux/ -xv bin/kubelet && touch staging/linux/bin/kubelet
tar: Failed to set default locale
x bin/kubelet
docker build -t k0sbuild.docker-image.containerd \
                --build-arg VERSION=1.5.5 \
                --build-arg BUILDIMAGE=golang:1.16-alpine \
                --build-arg BUILD_GO_TAGS="apparmor,selinux" \
                --build-arg BUILD_GO_CGO_ENABLED= \
                --build-arg BUILD_SHIM_GO_CGO_ENABLED=0 \
                --build-arg BUILD_GO_FLAGS= \
                --build-arg BUILD_GO_LDFLAGS= \
                --build-arg BUILD_GO_LDFLAGS_EXTRA="-w -s -extldflags=-static" \
                -f containerd/Dockerfile .
[+] Building 0.4s (12/12) FINISHED                                                                                                                                                                        
 => [internal] load build definition from Dockerfile                                                                                                                                                 0.0s
 => => transferring dockerfile: 1.04kB                                                                                                                                                               0.0s
 => [internal] load .dockerignore                                                                                                                                                                    0.0s
 => => transferring context: 34B                                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/golang:1.16-alpine                                                                                                                                0.3s
 => [build 1/7] FROM docker.io/library/golang:1.16-alpine@sha256:69d613a986635e0ee3f1b351131fa689b0dd3570ef9cab8e619a5f1fb939ec56                                                                    0.0s
 => CACHED [build 2/7] RUN apk upgrade -U -a && apk add  build-base git  btrfs-progs-dev btrfs-progs-static  protoc                                                                                  0.0s
 => CACHED [build 3/7] RUN mkdir -p /go/src/github.com/containerd/containerd                                                                                                                         0.0s
 => CACHED [build 4/7] RUN git clone -b v1.5.5 --depth=1 https://github.com/containerd/containerd.git /go/src/github.com/containerd/containerd                                                       0.0s
 => CACHED [build 5/7] WORKDIR /go/src/github.com/containerd/containerd                                                                                                                              0.0s
 => CACHED [build 6/7] RUN go version                                                                                                                                                                0.0s
 => CACHED [build 7/7] RUN make  CGO_ENABLED=  SHIM_CGO_ENABLED=0  GO_TAGS="-tags=apparmor,selinux"  COMMANDS="containerd containerd-shim containerd-shim-runc-v1 containerd-shim-runc-v2"  GO_BUIL  0.0s
 => CACHED [stage-1 1/1] COPY --from=build /go/src/github.com/containerd/containerd/bin/* /bin/                                                                                                      0.0s
 => exporting to image                                                                                                                                                                               0.0s
 => => exporting layers                                                                                                                                                                              0.0s
 => => writing image sha256:a37d848bfb323696b0ee7e0aa273344f682f2fc03c3985624d9c3ccb8c62d1c0                                                                                                         0.0s
 => => naming to docker.io/library/k0sbuild.docker-image.containerd                                                                                                                                  0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
touch .docker-image.containerd.stamp
docker create k0sbuild.docker-image.containerd > .container.containerd.tmp
mv .container.containerd.tmp .container.containerd
docker export $(cat .container.containerd) | tar -C staging/linux/ -xv bin/containerd && touch staging/linux/bin/containerd
tar: Failed to set default locale
x bin/containerd
docker export $(cat .container.containerd) | tar -C staging/linux/ -xv bin/containerd-shim && touch staging/linux/bin/containerd-shim
tar: Failed to set default locale
x bin/containerd-shim
docker export $(cat .container.containerd) | tar -C staging/linux/ -xv bin/containerd-shim-runc-v1 && touch staging/linux/bin/containerd-shim-runc-v1
tar: Failed to set default locale
x bin/containerd-shim-runc-v1
docker export $(cat .container.containerd) | tar -C staging/linux/ -xv bin/containerd-shim-runc-v2 && touch staging/linux/bin/containerd-shim-runc-v2
tar: Failed to set default locale
x bin/containerd-shim-runc-v2
docker export $(cat .container.kubernetes) | tar -C staging/linux/ -xv bin/kube-apiserver && touch staging/linux/bin/kube-apiserver
tar: Failed to set default locale
x bin/kube-apiserver
docker export $(cat .container.kubernetes) | tar -C staging/linux/ -xv bin/kube-scheduler && touch staging/linux/bin/kube-scheduler
tar: Failed to set default locale
x bin/kube-scheduler
docker export $(cat .container.kubernetes) | tar -C staging/linux/ -xv bin/kube-controller-manager && touch staging/linux/bin/kube-controller-manager
tar: Failed to set default locale
x bin/kube-controller-manager
docker build -t k0sbuild.docker-image.etcd \
                --build-arg VERSION=3.5.0 \
                --build-arg BUILDIMAGE=golang:1.16-alpine \
                --build-arg BUILD_GO_TAGS= \
                --build-arg BUILD_GO_CGO_ENABLED=0 \
                --build-arg BUILD_SHIM_GO_CGO_ENABLED= \
                --build-arg BUILD_GO_FLAGS= \
                --build-arg BUILD_GO_LDFLAGS="-w -s" \
                --build-arg BUILD_GO_LDFLAGS_EXTRA= \
                -f etcd/Dockerfile .
[+] Building 0.7s (11/11) FINISHED                                                                                                                                                                        
 => [internal] load build definition from Dockerfile                                                                                                                                                 0.0s
 => => transferring dockerfile: 766B                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                    0.0s
 => => transferring context: 34B                                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/golang:1.16-alpine                                                                                                                                0.6s
 => [build 1/6] FROM docker.io/library/golang:1.16-alpine@sha256:69d613a986635e0ee3f1b351131fa689b0dd3570ef9cab8e619a5f1fb939ec56                                                                    0.0s
 => CACHED [build 2/6] RUN apk add build-base git                                                                                                                                                    0.0s
 => CACHED [build 3/6] RUN cd / && git clone -b v3.5.0 --depth=1 https://github.com/etcd-io/etcd.git                                                                                                 0.0s
 => CACHED [build 4/6] WORKDIR /etcd/server                                                                                                                                                          0.0s
 => CACHED [build 5/6] RUN go version                                                                                                                                                                0.0s
 => CACHED [build 6/6] RUN CGO_ENABLED=0     go build           -installsuffix=cgo         -tags=""         -ldflags="-w -s  -X=go.etcd.io/etcd/api/v3/version.GitSHA=$(git rev-parse --short HEAD   0.0s
 => CACHED [stage-1 1/1] COPY --from=build /bin/etcd /bin/etcd                                                                                                                                       0.0s
 => exporting to image                                                                                                                                                                               0.0s
 => => exporting layers                                                                                                                                                                              0.0s
 => => writing image sha256:150cc534ae66817e609527621c6eea2ffe9ae511edbf28fbf4a1fb413db4e910                                                                                                         0.0s
 => => naming to docker.io/library/k0sbuild.docker-image.etcd                                                                                                                                        0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
touch .docker-image.etcd.stamp
docker create k0sbuild.docker-image.etcd > .container.etcd.tmp
mv .container.etcd.tmp .container.etcd
docker export $(cat .container.etcd) | tar -C staging/linux/ -xv bin/etcd && touch staging/linux/bin/etcd
tar: Failed to set default locale
x bin/etcd
docker build -t k0sbuild.docker-image.kine \
                --build-arg VERSION=0.7.1 \
                --build-arg BUILDIMAGE=golang:1.16-alpine \
                --build-arg BUILD_GO_TAGS= \
                --build-arg BUILD_GO_CGO_ENABLED= \
                --build-arg BUILD_SHIM_GO_CGO_ENABLED= \
                --build-arg BUILD_GO_FLAGS= \
                --build-arg BUILD_GO_LDFLAGS="-w -s" \
                --build-arg BUILD_GO_LDFLAGS_EXTRA="-extldflags=-static" \
                -f kine/Dockerfile .
[+] Building 0.4s (11/11) FINISHED                                                                                                                                                                        
 => [internal] load build definition from Dockerfile                                                                                                                                                 0.0s
 => => transferring dockerfile: 641B                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                    0.0s
 => => transferring context: 34B                                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/golang:1.16-alpine                                                                                                                                0.2s
 => [build 1/6] FROM docker.io/library/golang:1.16-alpine@sha256:69d613a986635e0ee3f1b351131fa689b0dd3570ef9cab8e619a5f1fb939ec56                                                                    0.0s
 => CACHED [build 2/6] RUN apk add build-base git                                                                                                                                                    0.0s
 => CACHED [build 3/6] RUN cd / && git clone -b v0.7.1 --depth=1 https://github.com/rancher/kine.git                                                                                                 0.0s
 => CACHED [build 4/6] WORKDIR /kine                                                                                                                                                                 0.0s
 => CACHED [build 5/6] RUN go version                                                                                                                                                                0.0s
 => CACHED [build 6/6] RUN CGO_ENABLED=     go build                  -tags=""         -ldflags="-w -s -extldflags=-static"         -o kine                                                          0.0s
 => CACHED [stage-1 1/1] COPY --from=build /kine/kine /bin/kine                                                                                                                                      0.0s
 => exporting to image                                                                                                                                                                               0.0s
 => => exporting layers                                                                                                                                                                              0.0s
 => => writing image sha256:bec37dd062d0d13dd6d6105f3fc24d12e12226abfcda3bc8286c8c3473d6771e                                                                                                         0.0s
 => => naming to docker.io/library/k0sbuild.docker-image.kine                                                                                                                                        0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
touch .docker-image.kine.stamp
docker create k0sbuild.docker-image.kine > .container.kine.tmp
mv .container.kine.tmp .container.kine
docker export $(cat .container.kine) | tar -C staging/linux/ -xv bin/kine && touch staging/linux/bin/kine
tar: Failed to set default locale
x bin/kine
docker build -t k0sbuild.docker-image.konnectivity \
                --build-arg VERSION=0.0.24 \
                --build-arg BUILDIMAGE=golang:1.16-alpine \
                --build-arg BUILD_GO_TAGS= \
                --build-arg BUILD_GO_CGO_ENABLED=0 \
                --build-arg BUILD_SHIM_GO_CGO_ENABLED= \
                --build-arg BUILD_GO_FLAGS="-a" \
                --build-arg BUILD_GO_LDFLAGS="-w -s" \
                --build-arg BUILD_GO_LDFLAGS_EXTRA="-extldflags=-static" \
                -f konnectivity/Dockerfile .
[+] Building 31.6s (9/10)                                                                                                                                                                                 
 => [internal] load build definition from Dockerfile                                                                                                                                                 0.0s
 => => transferring dockerfile: 904B                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                    0.0s
 => => transferring context: 34B                                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/golang:1.16-alpine                                                                                                                                0.2s
 => [build 1/6] FROM docker.io/library/golang:1.16-alpine@sha256:69d613a986635e0ee3f1b351131fa689b0dd3570ef9cab8e619a5f1fb939ec56                                                                    0.0s
 => CACHED [build 2/6] RUN apk add build-base git make                                                                                                                                               0.0s
 => CACHED [build 3/6] RUN git clone -b v0.0.24 --depth=1 https://github.com/kubernetes-sigs/apiserver-network-proxy.git /apiserver-network-proxy                                                    0.0s
 => CACHED [build 4/6] WORKDIR /apiserver-network-proxy                                                                                                                                              0.0s
 => CACHED [build 5/6] RUN go version                                                                                                                                                                0.0s
 => ERROR [build 6/6] RUN GO111MODULE=on go get github.com/golang/mock/mockgen@v1.4.4 &&     make gen &&     CGO_ENABLED=0     GOOS=linux     go build         -a         -tags=""         -ldflag  31.2s
------                                                                                                                                                                                                    
 > [build 6/6] RUN GO111MODULE=on go get github.com/golang/mock/mockgen@v1.4.4 &&     make gen &&     CGO_ENABLED=0     GOOS=linux     go build         -a         -tags=""         -ldflags="-w -s -extldflags=-static"         -o bin/proxy-server cmd/server/main.go:                                                                                                                                            
#9 2.182 go: downloading github.com/golang/mock v1.4.4                                                                                                                                                    
#9 2.308 go: downloading golang.org/x/tools v0.0.0-20210106214847-113979e3529a                                                                                                                            
#9 3.904 go: downloading golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1                                                                                                                          
#9 3.904 go: downloading golang.org/x/mod v0.3.0
#9 6.330 go mod download
#9 30.92 mkdir -p /go/src
#9 30.92 protoc -I . proto/agent/agent.proto --go_out=plugins=grpc:/go/src
#9 30.92 make: protoc: No such file or directory
#9 30.92 make: *** [Makefile:110: proto/agent/agent.pb.go] Error 127
------
executor failed running [/bin/sh -c GO111MODULE=on go get github.com/golang/mock/mockgen@v1.4.4 &&     make gen &&     CGO_ENABLED=${BUILD_GO_CGO_ENABLED}     GOOS=linux     go build         ${BUILD_GO_FLAGS}         -tags="${BUILD_GO_TAGS}"         -ldflags="${BUILD_GO_LDFLAGS} ${BUILD_GO_LDFLAGS_EXTRA}"         -o bin/proxy-server cmd/server/main.go]: exit code: 2
make[1]: *** [.docker-image.konnectivity.stamp] Error 1
make: *** [.bins.linux.stamp] Error 2
```

xinfengliu commented:
There should not appear the line `mkdir -p /go/src`
Please make sure all your docker images are updated and cleanup your docker builder cache then try again.

makhov commented:
Why shouldn't it? It exists in the make file:
https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/master/Makefile#L109

soider commented:
After running `docker system prune -a` I was able to reproduce build error during the `docker build` for konnectivity

The command I used:

```
make clean
docker system prune -a
docker build -t k0sbuild.docker-image.konnectivity \
                --build-arg VERSION=0.0.24 \
                --build-arg BUILDIMAGE=golang:1.16-alpine \
                --build-arg BUILD_GO_TAGS= \
                --build-arg BUILD_GO_CGO_ENABLED=0 \
                --build-arg BUILD_SHIM_GO_CGO_ENABLED= \
                --build-arg BUILD_GO_FLAGS="-a" \
                --build-arg BUILD_GO_LDFLAGS="-w -s" \
                --build-arg BUILD_GO_LDFLAGS_EXTRA="-extldflags=-static" \
                -f embedded-bins/konnectivity/Dockerfile .
```


BUT if I all the same using `make k0s` at the end, it works. I think the root cause is somewhere in the go build cache vilume shared between steps, might it be that protoc binary is downloaded by some other build and stored in the volume which is than just mounted. So it looks like the build is green because of some data race in a sense.

I'd recommend to manually run konnectivity build image with mounter volume (the same way as it goes for the build) and check if protoc is there.

makhov commented:
Thanks, @soider, I'll give it a try a bit later.

BTW, according to @xinfengliu's [log](https://github.com/k0sproject/k0s/pull/1143#issuecomment-927635504) `proto/agent/agent.pb.go` and `konnectivity-client/proto/client/client.pb.go` targets aren't triggered. 
Probably the files are generated by some of the previous builds.

Personally, I prefer not to rely on build order and keep images self-contained.

xinfengliu commented:
The [log](https://github.com/k0sproject/k0s/pull/1143#issuecomment-927635504) I previously shared was a separate build: `make .docker-image.konnectivity.stamp` . I even tried `--no-cache` for docker build. 

The reason is that: the `proto/agent/agent.pb.go` and `konnectivity-client/proto/client/client.pb.go` are already there after `git clone`.

makhov commented:
@xinfengliu yep, you're right. 
And looks like I found what is the difference: I have docker buildkit enabled (I use Docker for Mac)
`DOCKER_BUILDKIT=0 make .docker-image.konnectivity.stamp` works well, but 
`DOCKER_BUILDKIT=1 make .docker-image.konnectivity.stamp` fails always.

xinfengliu commented:
I tried both with and without `buildkit` actually, both were successful.
I guess it maybe related to buildkit cache in your failed build. How about using `--no-cache` to build?
```
=> CACHED [build 3/6] RUN git clone -b v0.0.24 --depth=1 https://github.com/kubernetes-sigs/apiserver-network-proxy.git /apiserver-network-proxy                                                    0.0s
```

makhov commented:
@xinfengliu 
I ran some tests and here are the quick results:
- `docker build --no-cache -t k0sbuild.docker-image.konnectivity` works
- `docker prune && docker build -t k0sbuild.docker-image.konnectivity` works
- `docker prune && make build` fails with the same error.


```
docker build -t k0sbuild.docker-image.konnectivity \
                --build-arg VERSION=0.0.24 \
                --build-arg BUILDIMAGE=golang:1.16-alpine \
                --build-arg BUILD_GO_TAGS= \
                --build-arg BUILD_GO_CGO_ENABLED=0 \
                --build-arg BUILD_SHIM_GO_CGO_ENABLED= \
                --build-arg BUILD_GO_FLAGS="-a" \
                --build-arg BUILD_GO_LDFLAGS="-w -s" \
                --build-arg BUILD_GO_LDFLAGS_EXTRA="-extldflags=-static" \
                -f konnectivity/Dockerfile .
[+] Building 61.6s (10/11)                                                                                                                                                                                
 => [internal] load build definition from Dockerfile                                                                                                                                                 0.0s
 => => transferring dockerfile: 909B                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                    0.0s
 => => transferring context: 34B                                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/golang:1.16-alpine                                                                                                                                1.9s
 => [auth] library/golang:pull token for registry-1.docker.io                                                                                                                                        0.0s
 => CACHED [build 1/6] FROM docker.io/library/golang:1.16-alpine@sha256:69d613a986635e0ee3f1b351131fa689b0dd3570ef9cab8e619a5f1fb939ec56                                                             0.0s
 => [build 2/6] RUN apk add build-base git make tree                                                                                                                                                23.6s
 => [build 3/6] RUN git clone -b v0.0.24 --depth=1 https://github.com/kubernetes-sigs/apiserver-network-proxy.git /apiserver-network-proxy                                                           1.6s 
 => [build 4/6] WORKDIR /apiserver-network-proxy                                                                                                                                                     0.0s 
 => [build 5/6] RUN go version                                                                                                                                                                       0.3s 
 => ERROR [build 6/6] RUN GO111MODULE=on go get github.com/golang/mock/mockgen@v1.4.4 &&     make gen &&     CGO_ENABLED=0     GOOS=linux     go build         -a         -tags=""         -ldflag  34.1s 
------                                                                                                                                                                                                    
 > [build 6/6] RUN GO111MODULE=on go get github.com/golang/mock/mockgen@v1.4.4 &&     make gen &&     CGO_ENABLED=0     GOOS=linux     go build         -a         -tags=""         -ldflags="-w -s -extldflags=-static"         -o bin/proxy-server cmd/server/main.go:                                                                                                                                            
#10 2.610 go: downloading github.com/golang/mock v1.4.4                                                                                                                                                   
#10 2.759 go: downloading golang.org/x/tools v0.0.0-20210106214847-113979e3529a                                                                                                                           
#10 4.188 go: downloading golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1                                                                                                                         
#10 4.189 go: downloading golang.org/x/mod v0.3.0
#10 6.222 go mod download
#10 33.72 mkdir -p /go/src
#10 33.72 protoc -I . konnectivity-client/proto/client/client.proto --go_out=plugins=grpc:/go/src
#10 33.72 make: protoc: No such file or directory
#10 33.72 make: *** [Makefile:104: konnectivity-client/proto/client/client.pb.go] Error 127
------
executor failed running [/bin/sh -c GO111MODULE=on go get github.com/golang/mock/mockgen@v1.4.4 &&     make gen &&     CGO_ENABLED=${BUILD_GO_CGO_ENABLED}     GOOS=linux     go build         ${BUILD_GO_FLAGS}         -tags="${BUILD_GO_TAGS}"         -ldflags="${BUILD_GO_LDFLAGS} ${BUILD_GO_LDFLAGS_EXTRA}"         -o bin/proxy-server cmd/server/main.go]: exit code: 2
make[1]: *** [.docker-image.konnectivity.stamp] Error 1
make: *** [.bins.linux.stamp] Error 2
```

makhov commented:
Sorry, by `docker prune` I meant a combination of `docker builder prune` and `docker system prune`. 

Current results show that it's most likely a bug somewhere in buildkit. I'll try to dig deeper, but it will take some time. 
WDYT, should we close the PR and open an issue instead?

```
❯ docker version
Client:
 Cloud integration: 1.0.17
 Version:           20.10.8
 API version:       1.41
 Go version:        go1.16.6
 Git commit:        3967b7d
 Built:             Fri Jul 30 19:55:20 2021
 OS/Arch:           darwin/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.8
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.6
  Git commit:       75249d8
  Built:            Fri Jul 30 19:52:10 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.9
  GitCommit:        e25210fe30a0a703442421b0f60afac609f950a3
 runc:
  Version:          1.0.1
  GitCommit:        v1.0.1-0-g4144b63
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

xinfengliu commented:
I feel it's better to close this PR and open an new issue. Just adding installation of `protoc` in the Dockerfile may be not OK due to https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/110 .  

It's strange that in your environment `proto/agent/agent.pb.go` and `konnectivity-client/proto/client/client.pb.go` were being rebuilt. Unless they do not exist or the correspondent `.proto` files are touched, they should not be rebuilt.

makhov commented:
Yeah, totally agree. Will create an issue a bit. Thanks, @soider @xinfengliu!
